stor/dx.py

```diff
@@ -114,71 +157,672 @@ def temp_url(self, lifetime=300, filename=None):
+
+    def _clone(self, dest):
+        """Clones the data object into the destination path.
+        The original file is retained.
+
+        For example, assume the following file hierarchy::
+
+            dxProject/
+            - a/
+            - - 1.txt
+
+            anotherDxProject/
+
+        Doing a clone of ``1.txt`` to a new destination of ``b.txt`` is
+        performed with::
+
+            Path('dx://dxProject:/a/1.txt')._clone(Path('dx://anotherDxProject/b.txt'))
+
+        The end result for anotherDxProject looks like::
+
+            anotherDxProject/
+            - b.txt
+
+        If the dest already exists as a directory, the file is moved inside it, retaining
+        its original name. Thus, if the original project structure is
+
+            dxProject/
+            - a/
+            - - 1.txt
+
+            anotherDxProject/
+            - b.txt/
+
+        Performing clone with following command::
+
+            Path('dx://dxProject:/a/1.txt')._clone(Path('dx://anotherDxProject/b.txt'))
+
+        Will yield the resulting structure to be:
+
+            anotherDxProject/
+            - b.txt/
+            - - 1.txt
+
+        Args:
+            dest (Path): The destination file/folder path in a different project
+
+        Raises:
+            ValueError: If attempting to clone a project
+            DNAnexusError: If cloning within same project
+            TargetExistsError: When all possible destinations for source file already exist
+        """
+        if not self.resource:
+            raise ValueError('Cannot clone project ({})'.format(self))
+        if dest.canonical_project == self.canonical_project:
+            raise DNAnexusError('Cannot clone within same project')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        dest_is_dir = dest.isdir()
+        if dest_is_dir:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise TargetExistsError(
+                    'Destination path ({}) already exists, will not cause '
+                    'duplicate file objects to exist. Remove the original first'
+                    .format(new_dest))
+            folder_dest = dest
         else:
-            return self.stat()['size']
+            if dest.isfile():
+                raise TargetExistsError(
+                    'Destination path ({}) already exists, will not cause '
+                    'duplicate file objects to exist. Remove the original first'
+                    .format(dest)
+                )
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        with _propagate_dx_exceptions():
+            new_file_h = file_handler.clone(project=dest.canonical_project,
+                                            folder=folder_dest.resource or '/')
+            if not dest_is_dir and not utils.has_trailing_slash(dest):
```

Comment:
No, the use case is not from dxpy's side but from the _clone functionality side. We only need to rename if the final destination's name is not the same. Say we execute `Path('dx://project:/file').copy('dx://project2:/folder/')` where `project2:/folder/` doesn't exist. Then dest_is_dir is False (because it doesn't already exist as a directory), but we don't need to rename in this case because the original filename is used for copying to `Path('dx://project:/folder/file')`